File: backend/repositories/models/class_model.py
from sqlalchemy import Column, Integer, String, Text, Boolean, ForeignKey, TIMESTAMP
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from repositories.database import Base


class Class(Base):
    """
    SQLAlchemy model for classes table
    Represents a class/course taught by a teacher
    """
    __tablename__ = "classes"

    id = Column(Integer, primary_key=True, index=True)
    teacher_id = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    class_name = Column(String(100), nullable=False)
    class_code = Column(String(20), unique=True, nullable=False, index=True)
    description = Column(Text, nullable=True)
    created_at = Column(TIMESTAMP(timezone=True), nullable=False, server_default=func.now())
    is_active = Column(Boolean, nullable=False, default=True)

    # Relationships
    teacher = relationship("User", back_populates="classes", lazy="selectin")
    assignments = relationship("Assignment", back_populates="class_obj", cascade="all, delete-orphan", lazy="selectin")
    enrollments = relationship("Enrollment", back_populates="class_obj", cascade="all, delete-orphan", lazy="selectin")

    def __repr__(self):
        return f"<Class(id={self.id}, name='{self.class_name}', code='{self.class_code}')>"

File: backend-ts/src/models/class.model.ts
import {
    pgTable,
    serial,
    integer,
    varchar,
    text,
    boolean,
    timestamp,
} from 'drizzle-orm/pg-core';
import { relations } from 'drizzle-orm';
import { users } from './user.model.js';

/** Classes table - represents courses taught by teachers */
export const classes = pgTable('classes', {
    id: serial('id').primaryKey(),
    teacherId: integer('teacher_id').notNull().references(() => users.id, { onDelete: 'cascade' }),
    className: varchar('class_name', { length: 100 }).notNull(),
    classCode: varchar('class_code', { length: 20 }).unique().notNull(),
    description: text('description'),
    createdAt: timestamp('created_at', { withTimezone: true }).defaultNow().notNull(),
    isActive: boolean('is_active').default(true).notNull(),
});

/** Class relations */
export const classesRelations = relations(classes, ({ one, many }) => ({
    teacher: one(users, {
        fields: [classes.teacherId],
        references: [users.id],
    }),
    assignments: many(assignments),
    enrollments: many(enrollments),
}));

/** Type definitions for Class */
export type Class = typeof classes.$inferSelect;
export type NewClass = typeof classes.$inferInsert;

// Import related tables
import { assignments } from './assignment.model.js';
import { enrollments } from './enrollment.model.js';
